pkg/otel/common/schema/transform/dictionary.go

```diff
@@ -156,23 +171,32 @@ func (t *DictionaryField) updateIndexType(stats *stats.RecordBuilderStats) {
 		return
 	}
 
+	prevIndexType := t.IndexType()
 	currentIndex := t.currentIndex
 
 	for t.currentIndex < len(t.indexTypes) && t.cardinality > t.indexMaxCard[t.currentIndex] {
 		t.currentIndex++
 	}
 	if t.currentIndex >= len(t.indexTypes) {
-		t.indexTypes = nil
-		t.indexMaxCard = nil
-		t.currentIndex = 0
-		t.schemaUpdateRequest.Inc()
-		t.events.DictionariesWithOverflow[t.path] = true
-		stats.DictionaryOverflowDetected++
+		ratio := float64(t.cardinality) / float64(t.cumulativeTotal)
+		if ratio < t.config.ResetThreshold {
+			t.currentIndex = len(t.indexTypes) - 1
+			t.schemaUpdateRequest.Inc(&update.DictionaryResetEvent{FieldName: t.path, IndexType: t.IndexType(), Cardinality: t.cardinality, Total: t.cumulativeTotal})
+			t.cumulativeTotal = 0
```

Comment:
Looks good. Now would be a good time to visit the top of this file and comment these three fields:
```
	indexMaxCard []uint64
	indexTypes   []arrow.DataType
	currentIndex int
```
(I went to study the code to be sure I'm following, thanks.)

Reply:
Added the following comments.

```go
	// The following index defines which index type and max cardinality is
	// currently used. Each time the cardinality is updated, the index is
	// index type is reevaluated to determine if the new cardinality exceeds
	// the max cardinality of the current index type. If it does, the index
	// is incremented until the max cardinality of the new index type is
	// greater than the current cardinality.
	// When the cardinality exceeds the max cardinality of the last index
	// type, the dictionary is either reset or overflowed depending on the
	// ratio between the cardinality and the cumulative total.
	currentIndex int
	// The different max cardinalities used for this column. This slice is
	// initialized based on the min and max cardinalities defined in the
	// dictionary configuration.
	indexMaxCard []uint64
	// The different index types used for this column aligned with the
	// indexMaxCard slice.
	indexTypes []arrow.DataType
```